api/src/statistics/members_of_interest.py

```diff
@@ -0,0 +1,267 @@
+) -> List[Tuple[Member, float]]:
+    product_ids = [product.id for product in products]
+    stmt = (
+        select(Member, func.sum(TransactionContent.amount).label("amount"))
+        .where(Product.id.in_(product_ids))
+        .where(Transaction.created_at >= start if start is not None else sqlalchemy.cast(True, sqlalchemy.Boolean))
+        .where(Transaction.created_at < end if end is not None else sqlalchemy.cast(True, sqlalchemy.Boolean))
+        .order_by(func.sum(TransactionContent.amount).desc())
+        .limit(limit)
+        .group_by(Member.member_id)
+        .join(Transaction, Member.member_id == Transaction.member_id)
+        .join(TransactionContent, Transaction.id == TransactionContent.transaction_id)
+        .join(TransactionContent.product)
+    )
+    return [(member, amount) for (member, amount) in db_session.execute(stmt)]
+
+
+def days_with_visits(
+    start: Optional[datetime], end: Optional[datetime], limit: Optional[int] = 30
+) -> List[Tuple[Member, int]]:
+    unique_days = func.count(distinct(func.DATE(PhysicalAccessEntry.invoked_at)))
+    stmt = (
+        select(Member, unique_days.label("days_with_visits"))
+        .where(
+            PhysicalAccessEntry.created_at >= start if start is not None else sqlalchemy.cast(True, sqlalchemy.Boolean)
+        )
+        .where(PhysicalAccessEntry.created_at < end if end is not None else sqlalchemy.cast(True, sqlalchemy.Boolean))
+        .having(unique_days > 0)
+        .limit(limit)
+        .order_by(desc(unique_days))
+        .group_by(Member.member_id)
+        .join(PhysicalAccessEntry, Member.member_id == PhysicalAccessEntry.member_id)
+    )
+    return [(member, count) for (member, count) in db_session.execute(stmt)]
+
+
+def percentiles(members: List[Tuple[Member, int]]) -> Dict[Member, Tuple[float, int]]:
+    n = len(members)
+    result: Dict[Member, Tuple[float, int]] = {}
+    for i, (member, value) in enumerate(members):
+        percentile = (i + 1) / n
+        result[member] = (percentile, value)
+    return result
+
+
+@dataclass
+class ActivityGroup:
+    name: str
+    scores: Dict[str, float]
+
+
+@dataclass
+class MemberScore(DataClassJsonMixin):
+    member_id: int
+    member_number: int
+    firstname: str
+    lastname: str | None
+    activity_percentile: float
+    score: float
+
+
+@dataclass
+class ActivityGroupStatistics(DataClassJsonMixin):
+    name: str
+    products: List[Tuple[int, str, float]]
+    doors: List[Tuple[str, str, float]]
+    member_scores: List[MemberScore]
+
+
+@dataclass
+class MembersOfInterestResponse(DataClassJsonMixin):
+    groups: List[ActivityGroupStatistics]
+
+
+def members_of_interest(start: Optional[datetime], end: Optional[datetime]) -> MembersOfInterestResponse:
+    member_activity = days_with_visits(start, end, limit=None)
+    member_percentiles = percentiles(member_activity)
+
+    # TODO: This should ideally be moved to a config file
```

Comment:
Agreed. You can create an issue and refer to the "Generalization" project (https://github.com/orgs/makerspace/projects/4)

Reply:
https://github.com/makerspace/makeradmin/issues/626